src/quickcheck.rs

```diff
@@ -0,0 +1,264 @@
+            0,
+            max(
+                1,
+                min(
+                    u32::try_from(g.size()).unwrap_or(u32::max_value()),
+                    1_000_000_000,
+                ),
+            ),
+        );
+        Self::from_hms_nanos_unchecked(h, m, s, ns)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let orig_h = self.hour;
+        let orig_m = self.minute;
+        let orig_s = self.second;
+        let orig_nanos = self.nanosecond;
+
+        let shrunk_hour = orig_h
+            .shrink()
+            .map(move |h| Self::from_hms_nanos_unchecked(h, orig_m, orig_s, orig_nanos));
+        let shrunk_minute = orig_m
+            .shrink()
+            .map(move |m| Self::from_hms_nanos_unchecked(orig_h, m, orig_s, orig_nanos));
+        let shrunk_second = orig_s
+            .shrink()
+            .map(move |s| Self::from_hms_nanos_unchecked(orig_h, orig_m, s, orig_nanos));
+        let shrunk_nanos = orig_nanos
+            .shrink()
+            .map(move |nanos| Self::from_hms_nanos_unchecked(orig_h, orig_m, orig_s, nanos));
+
+        Box::new(
+            shrunk_hour
+                .chain(shrunk_minute)
+                .chain(shrunk_second)
+                .chain(shrunk_nanos),
+        )
+    }
+}
+
+impl Arbitrary for PrimitiveDateTime {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let d = Date::arbitrary(g);
+        let t = Time::arbitrary(g);
+        Self::new(d, t)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let orig_d = self.date;
+        let orig_t = self.time;
+
+        let shrunk_date = orig_d.shrink().map(move |d| Self::new(d, orig_t));
+        let shrunk_time = orig_t.shrink().map(move |t| Self::new(orig_d, t));
+
+        Box::new(shrunk_date.chain(shrunk_time))
+    }
+}
+
+impl Arbitrary for UtcOffset {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let size = min(
+            i32::try_from(g.size()).unwrap_or(i32::max_value()),
+            60 * 60 * 24,
+        );
+        let o = g.gen_range(-(size - 1), size);
+        Self::seconds_unchecked(o)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        Box::new(self.as_seconds().shrink().map(Self::seconds_unchecked))
+    }
+}
+
+impl Arbitrary for OffsetDateTime {
+    fn arbitrary<G>(g: &mut G) -> Self
+    where
+        G: Gen,
+    {
+        let dt = PrimitiveDateTime::arbitrary(g);
+        let o = UtcOffset::arbitrary(g);
+        dt.assume_offset(o)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let orig_dt = PrimitiveDateTime::new(self.date(), self.time());
+        let orig_o = self.offset();
+
+        let shrunk_datetime = orig_dt.shrink().map(move |dt| dt.assume_offset(orig_o));
+        let shrunk_offset = orig_o.shrink().map(move |o| orig_dt.assume_offset(o));
+
+        Box::new(shrunk_datetime.chain(shrunk_offset))
+    }
+}
```

Comment:
For `OffsetDateTime` — what would a user be expecting to happen during shrinking? I think (read: I haven't checked this) that this effectively shrinks the date, time, and offset, in that order. As the offset affects the calculated date/time, is this what a user would expect? I believe so, but I'm not sure :thinking:

Reply:
Yes, I believe it happens in that order - at least I had some issues in another project while implementing shrinking for a `BigRational` type. I had first shrunk the numerator and then the denominator (or maybe vice versa), which caused some issues when one of the two ended up many orders of magnitude greater than the other; reworking the shrinker to effectively alternate between them solved that issue.

My thinking here was to shrink the components in descending order of significance to traverse the value space efficiently, but I'm honestly not sure what a user would expect "shrinking a date" to do exactly. But I guess what matters most is the purpose of shrinking: to find minimal counterexamples. In that light I probably would as a user like to know if the particular offset is significant for the test failure (because it would be shrunk to zero if not). Ultimately I don't think the shrinking order matters much - you'll want the shrinker to generate likely continued failures first, but I don't think any of the components is more or less likely to be significant for any particular test failure.

Maybe it _would_ be more efficient to only generate one shrunk value per component per shrink step, though, so the shrinker doesn't have to run through N shrunk years and N shrunk days in each step before getting to the shrunk offsets, though. That would ultimately depend on the behaviour of `i32::shrink` etc., but I guess we could just limit that using `Iterator::take` anyway.

Reply:
I see your thumbs up, does that mean you're happy with this as is or would you like me to limit the iterator sizes like I said in the last paragraph? :slightly_smiling_face:

Reply:
Yeah, it's fine as-is.